File: Data Structure/linkedList.js
const LinkedList = (function () {
  function LinkedList() {
    this.length = 0;
    this.head = null;
  }
  function Node(data) {
    this.data = data;
    this.next = null;
  }
  LinkedList.prototype.add = function (value) {
    var node = new Node(value);
    var current = this.head;
    if (!current) {
      this.head = node;
      this.length++;

      return node;
    } else {
      while (current.next) { // 마지막 노드를 찾고.
        current = current.next;
      }
      current.next = node; // 마지막 위치에 노드 추가
      this.length++;
      return node;
    }

  };
  LinkedList.prototype.search = function (position) {
    var current = this.head;
    var count = 0;
    while (count < position) {
      current = current.next;
      count++;
    }
    return current.data;
  };
  LinkedList.prototype.remove = function (position) {
    var current = this.head;
    var before, remove, count = 0;
    if (position === 0) {
      remove = this.head;
      this.head = this.head.next;
      this.length--;
      return remove;
    }

    while (count < position) {
      before = current;
      count++;
      current = current.next;
    }
    remove = current;
    before.next = remove.next;
    this.length--;
    return remove;
  };
  return LinkedList;
})();

const list = new LinkedList();
list.add(1);
list.add(2);
list.add(3);
console.log(list.length);
console.log(list.search(0));
console.log(list.search(2));

list.remove(1);
console.log(list.length);
console.log(list.search(1));

File: Java/whiteship-java-study/LinkedList.java
class ListNode {
  private int value;
  private ListNode next;

  public ListNode() {}

  public ListNode(int value) {
    this.value = value;
    this.next = null;
  }

  public static ListNode add(ListNode head, ListNode nodeToAdd, int position) {
    if (position <= 0) {
      throw new IllegalArgumentException("position should be positive.");
    }

    if (head.next == null) {
      head.next = nodeToAdd;
      nodeToAdd.next = null;
      return nodeToAdd;
    }

    ListNode currentListNode = head;
    for (int i = 0; i < position - 1; i++) {
      currentListNode = currentListNode.next;
    }

    nodeToAdd.next = currentListNode.next;
    currentListNode.next = nodeToAdd;

    return currentListNode;
  }

  public static void getAll(ListNode head) {
    ListNode currentListNode = head;
    int position = 0;
    while (currentListNode != null) {
      System.out.println("value: " + currentListNode.value + ", position: " + position);
      currentListNode = currentListNode.next;
      position++;
    }
    System.out.println();
  }

  public static ListNode remove(ListNode head, int positionToRemove) {
    if (positionToRemove <= 0) {
      throw new IllegalArgumentException("position should be positive.");
    }

    ListNode currentListNode = head;
    for (int i = 0; i < positionToRemove - 1; i++) {
      currentListNode = currentListNode.next;
    }

    if (currentListNode.next == null) {
      throw new IndexOutOfBoundsException("empty linkedlist");
    }
    currentListNode.next = currentListNode.next.next;
    return currentListNode;
  }

  public static boolean contains(ListNode head, ListNode nodeToCheck) {
     ListNode currentListNode = head;
     while(currentListNode.next != nodeToCheck) {
       if (currentListNode.next == null) {
         return false;
       }
       currentListNode = currentListNode.next;
     }
     return true;
  }
}

public class LinkedList {
  public static void main(String[] args) {
    ListNode head = new ListNode();
    ListNode first = new ListNode(100);
    ListNode last = new ListNode(100);

    ListNode.add(head, first, 1);
    ListNode.add(head, new ListNode(1), 1);
    ListNode.add(head, new ListNode(2), 1);
    ListNode.add(head, new ListNode(3), 1);
    ListNode.add(head, new ListNode(4), 1);
    ListNode.add(head, new ListNode(5), 1);
    ListNode.add(head, last, 1);

    ListNode.getAll(head);
    ListNode.remove(head, 1);
    ListNode.remove(head, 1);
    ListNode.remove(head, 1);
    ListNode.getAll(head);

    System.out.println(ListNode.contains(head, first));
    System.out.println(ListNode.contains(head, last));
  }
}
